### PR TITLE
Fix dashboard port conflict message and allow starting on random port

### PR DESCRIPTION
**Pull request by pravinpushkar**
Signed-off-by: Pravin Pushkar <ppushkar@microsoft.com>

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #1188

### CODE REVIEW

File: cmd/dashboard.go [hunk 1]
```diff
@@ -99,6 +102,12 @@ dapr dashboard -k -p 0
 			os.Exit(1)
 		}
 
+		if err := utils.CheckIfPortAvailable(dashboardLocalPort); err != nil {
+			print.FailureStatusEvent(os.Stderr, "Please select a different port with %q flag: %s", "-p", err)
```
Comment (shubham1172) [inline cmd/dashboard.go:106]: Will this error message inform the user that the port is busy?

Comment (shubham1172) [reply cmd/dashboard.go:106]: Do we need "%q" with "-p", can we use the flag in string directly?

File: utils/utils.go [hunk 2]
```diff
@@ -266,6 +266,15 @@ func IsAddressLegal(address string) bool {
 	return isLegal
 }
 
+func CheckIfPortAvailable(port int) error {
```
Comment (shubham1172) [inline utils/utils.go:269]: Please add doc comment, can we also add tests?

**Review by shubham1172 — CHANGES_REQUESTED** [review]

File: utils/utils.go (hunk 2 above)
Comment (pravinpushkar) [reply utils/utils.go:269]: I think for test we will have to depend on some random port which might be occupied in some cases and test will be error prone. thoughts?
Also, this method just calls the `net.Listen` and does not have much business logic.

File: cmd/dashboard.go (hunk 1 above)
Comment (pravinpushkar) [reply cmd/dashboard.go:106]: > Will this error message inform the user that the port is busy?

yes.

> Do we need "%q" with "-p", can we use the flag in string directly?

yes, we will have to use %q here.

File: utils/utils.go (hunk 2 above)
Comment (shubham1172) [reply utils/utils.go:270]: We can `net.Listen` on the same port before calling this function to test. The port can be fixed or generated from freeport pkg. 

Ok with skipping the test, but it's good to have. /cc @mukundansundar

**Review by shubham1172 — APPROVED** [review]

File: utils/utils.go (hunk 2 above)
Comment (shubham1172) [reply utils/utils.go:270]: As discussed, this test can be tracked separately.

**Review by mukundansundar — APPROVED** [review]

**Comment by pravinpushkar** [comment]
Below is the output - 
@mukundansundar @shubham1172 

#### Valid port/ default port is unoccupied - 

➜  cli git:(dashboard) ldapr dashboard
Dapr Dashboard running on http://localhost:8080

#### When default port is occupied - 
➜  ~ ldapr dashboard
❌  Please select a different port with "-p" flag: listen tcp :8080: bind: address already in use
ℹ️  You can also use port 0 to select a random free port.

#### When passed port is 0 - 
➜  ~ ldapr dashboard -p 0
Dapr Dashboard running on http://localhost:50190

File: utils/utils.go (hunk 2 above)
Comment (pravinpushkar) [reply utils/utils.go:270]: There is no E2E for dashboard. We can try to add one and cover this there.
tracking issue - https://github.com/dapr/cli/issues/1192

Comment (shubham1172) [reply utils/utils.go:270]: Thanks @pravinpushkar

**Pull request merged**